[4-9단계 - 방탈출 예약 관리] 멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

    [멍구 프롤로그](https://prolog.techcourse.co.kr/studylogs/4166)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

Repository에서 Reservation, ReservationTime 등의 객체를 받아 getter로 호출하는 부분이 다수 존재합니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

- 레이어드 아키텍처의 적용이 잘 되었는지
- 유효성 검사의 위치가 적절한지, 통일이 필요한지
    Reservation의 경우 다른 도메인(ReservationTime)까지 생성하여 조립하는 등 복잡도가 있고, ReservationTime은 비교적 간단합니다. 따라서 다르게 구현하였는데, 구현 스타일을 통일해야 할까요?

    - Reservation: Dto ↔ Domain Mapper 존재 / Service에서 날짜형식 검증
    - ReservationTime: Dto 내부에서 ↔ Domain 변환 / Dto 내부 변환로직 내에서 time 형식 검증

참고: 요구사항에 따라 4-9단계 MissionStepTest는 리팩터링 하지 않았습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

yuyoungrhee committed:
refactor(repository): 초기 데이터 개수와 index 시작 위치를 일치시키도록 변경

- 기존 3 하드코딩에서 데이터size를 계산하여 index를 초기화하는 방향으로 수정

yuyoungrhee committed:
chore: 4단계 - 데이터베이스 적용을 위한 환경 구성

yuyoungrhee committed:
refacor: 5단계 - 예약 조회 api를 h2 데이터베이스를 활용하도록 전환

- findAll()만 전환, 나머지는 주석 처리
- date, time이 DB에서 예약어이므로 의미있는 컬럼명으로 변경 (reservation_date, reservation_time)

yuyoungrhee committed:
refactor: 6단계 - 예약 추가/삭제 api를 h2 데이터베이스를 활용하도록 전환

- Reservation 정적펙터리메서드 이름 가독성 개선
- 데이터베이스 테이블 생성, 초기 데이터 추가 sql문 구현
- Repository의 메서드명 개선(add -> save)

yuyoungrhee committed:
refactor: setObject 대신 setDate와 setTime 사용으로 안정성 개선

- SQL 타입 명시를 명확하게 함, 가독성과 안정성 개선

yuyoungrhee committed:
feat: 7단계 - 시간 추가 api 구현

- reservation_time 초기화 구문 정의
- ReservationTime 관련 도메인, 레파지토리, Dto 구현

yuyoungrhee committed:
feat: 7단계 - 시간 조회 api 구현

yuyoungrhee committed:
feat: 7단계 - 시간 삭제 api 구현

yuyoungrhee committed:
feat: 8단계 - 예약과 예약 시간을 연동, 예약페이지에서 시간 선택 가능하도록 구현

- reservation 테이블에 time_id라는 reservation_time의 외래키 추가
- ReservationRepository가 time_id에 따라 정상동작하도록 수정
- ReservationRequest, ReservationResponse 형식 변경을 반영하여 수정
- 예약 페이지 html을 reservation-legacy -> reservation으로 변경

yuyoungrhee committed:
feat(service): 날짜 형식에 대한 유효성 검증 추가

yuyoungrhee committed:
test(service): ReservationService에 대한 테스트 추가

yuyoungrhee committed:
test(mapper): ReservationMapper 테스트 추가

- ReservationMapper의 상수 선언 위치 이동

yuyoungrhee committed:
refactor(TimeRequest): 상황에 맞는 예외로 변경

- parseStartAt()에서 발생하는 예외는 startAt이라는 외부값에 의존하므로 ArgumentException을 발생시키도록 변경

yuyoungrhee committed:
test(dto): TimeRequest/TimeResponse 테스트 추가

yuyoungrhee committed:
refactor(service): 9단계 - TimeService 추가 구현으로 계층화

yuyoungrhee committed:
refactor(repository): RowMapper 상수화

yuyoungrhee committed:
refactor(repository): SimpleJdbcInsert를 사용해 insert하도록 수정

yuyoungrhee committed:
fix(JdbcTimeRepository): findById()에서 id를 못 찾았을 때 예외 처리 추가

yuyoungrhee committed:
chore: model -> domain 패키지명 변경

yuyoungrhee committed:
feat(dto): request에 유효성 검사 추가

yuyoungrhee commented on src/main/java/roomescape/dto/request/ReservationRequest.java:
```diff
@@ -1,8 +1,19 @@
 package roomescape.dto.request;
 
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Pattern;
+
 public record ReservationRequest(
+
+        @NotBlank(message = "날짜는 필수입니다.")
+        @Pattern(regexp = "\\d{4}-\\d{2}-\\d{2}")
         String date,
+
+        @NotBlank
         String name,
-        String time
+
+        @NotNull
+        Long timeId
```
dto에도 검증을 추가했습니다. 

Dto와 Service 양쪽에 유효성 검증이 존재하는 이유는 다음과 같습니다.

**Request Dto의 유효성 검증**
→ 요청 값이 형식적으로 유효한가? (빈 값, 정규식 포맷 등)

**Service 내부의 유효성 검증**
→ 요청 값이 도메인에서 사용 가능한 값인가?
예) "25:00" 같은 잘못된 시간, "2월 30일" 같은 존재하지 않는 날짜

dusdn1702 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -1,47 +1,40 @@
-    public ReservationResponse addReservation(@RequestBody ReservationRequest request) {
-        Long id = repository.add(ReservationMapper.toDomain(request));
-        return ReservationMapper.toDto(repository.findById(id));
+    public ReservationResponse addReservation(@Valid @RequestBody ReservationRequest request) {
```
`@Valid`를 이용해 요청에 대한 유효성을 검증하신 이유가 궁금해요!

yuyoungrhee replied:
Jakatra Bean Validation 라이브러리를 통한 유효성 검증의 존재를 알고 있었습니다. 

여기서 dto에서 생성자를 이용해서 생성이 될 때 유효성 검사를 하는 방법도 고려했었습니다. 

두 방식을 비교해봤을 때, 어노테이션을 이용한 검증이 충분히 가독성 있으면서 코드의 복잡도를 줄여준다고 생각해서 적용했습니다.

이렇게 적용한 유효성 검증은 `@Valid`를 통해 컨트롤러 메서드 파라미터에서 자동으로 수행되어 RequestDto가 생성되는 시점에 검증이 함께 수행되게 됩니다!

dusdn1702 commented on src/main/java/roomescape/dto/request/TimeRequest.java:
```diff
@@ -0,0 +1,16 @@
+        @Pattern(regexp = "\\d{2}:\\d{2}")
+        String startAt
+) {
+    public ReservationTime toDomain() {
```
> 유효성 검사의 위치가 적절한지, 통일이 필요한지

위와 같이 말씀을 주셨는데, 변환 로직에 일관성 없이 다른 규칙을 가진다면 어떤 문제가 발생할 수 있을까요?
이렇게 상황에 맞춰 구현을 했을 때엔 어떤 장점이 있었나요?

yuyoungrhee replied:
장점
- 필요 이상의 복잡함을 만들지 않는다. (딱 필요한 때만 분리)

단점 및 문제점
- 협업자가 혼란스러울 수 있다. 
- ex) 변환 로직의 위치를 찾는데 어려움을 겪을 수 있다
- 추후 다른 도메인을 추가했을 때 dto 내부에 두어야하는지 Mapper로 분리해야하는지 등 선택에 어려움이 있을 수 있다

코드를 구현한 입장에서는 합리적인 선택일 수 있어도, 다른 크루들의 시선으로 피드백 받으면서 상당히 혼란스러울 수 있는 여지가 있는 것 같다고 판단했습니다. 따라서 Time도 Mapper를 사용해서 변환하도록 선택했습니다. 

샐리는 코드를 처음 보면서 혼란스러웠나요?
아니면 Dto는 어떻게 변환되어도 상관없다고 생각하셨나요? 
샐리의 시선도 궁금합니다.

dusdn1702 replied:
저는 모든 코드는 일관성을 갖는게 중요하다고 생각해요.
협업을 하다보면 멍구의 스타일은 A에서 변환하는 것이니 다른 객체도 A에서 변환하겠지. 라는 기대를 가지며 분석을 하게 되는데, B에서 변환을 하고 있으면 또 다른 객체는 어디서 변환할지 몰라 모든 객체를 찾느라 시간이 더욱 소요되고 혼란이 가중되죠.
그래서 컨벤션을 가지고 협업을 하고, 그에 맞추는 것이 중요하다고 봐요.
지금은 단순한 프로젝트이지만 이러한 코드들이 쌓여서 놓치기 쉬운 부분들이 되고, 유지보수에 어려움을 준다고 생각해요.

yuyoungrhee replied:
매우 동의합니다! 구현하는 입장에서만 바라보고, 이 코드가 누구에게 읽히고 유지보수하는지는 상관하지 않았던 것 같아요. 설계에 대해서 결정을 내릴 때 조금 더 넓은 시야로 바라보게 되었습니다. 좋은 답변 감사해요😀

dusdn1702 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -1,50 +1,12 @@
 
-@Component
-public class ReservationRepository {
+public interface ReservationRepository {
```
repository를 interface화 하신 이유가 무엇인가요?

yuyoungrhee replied:
Fake Repository를 이용한 Service 계층 테스트를 위해서 interface화 했습니다. 

프로덕션 코드를 위한 분리는 아니었지만, 자연스럽게 Service는 하위 모듈 의존하지 않고 추상에 의존해야한다는 원칙이 지켜졌기 때문에 허용 가능한 선택이라고 생각합니다.

dusdn1702 commented on src/main/java/roomescape/repository/JdbcReservationRepository.java:
```diff
@@ -0,0 +1,72 @@
+import roomescape.domain.ReservationTime;
+
+@Repository
+public class JdbcReservationRepository implements ReservationRepository {
```
`Jdbc` 라는 이름의 이유가 무엇인가요?

yuyoungrhee replied:
이 구현체는 JDBC 기반이라는 것을 명시하는 의미로 사용했습니다. 너무 구체적인 기술을 명시하는 네이밍이라 고민이 되긴 하네요...

Jdbc~~Repository는 JDBC 기반 구현체이기 때문에 내부적으로는 Dao 패턴의 구현 책임을 가집니다. Dao로 네이밍을 변경하여 Repository를 "구현한다"의 의미를 가져가는 게 좋을까요?

dusdn1702 replied:
Dao 패턴은 어떠한 의미라서 Repository보다 낫다고 보시나요? 
기술에 의존적인 repository이니 jdbc도 괜찮다고 볼 수 있을 것 같아요.
다만, repository가 프로덕션 코드에서 하나이기 때문에 그 명명에 의미가 있을지는 생각해보면 좋겠어요.

yuyoungrhee replied:
이해 했습니다! 그런데 Repository가 인터페이스가 되면서 그것을 구현하는 구현체이므로 특징이나 기술적인 의미를 나타내는 이름이 더 좋을 거 같습니다. "구현체"라는 의미 + Jdbc를 활용하여 Repository를 구현하고있다는 의미를 모두 주고 싶었기에 명명을 해주었던 것 같습니다.

그러나 샐리가 말해준 것처럼 Repository는 프로덕션 코드에서 하나이기 때문에, 만약 혼자 하는 프로그래밍이 아니라면, 이 부분은 정해지는 대로 따라갈 것 같긴 합니다!

dusdn1702 commented on src/main/java/roomescape/repository/JdbcReservationRepository.java:
```diff
@@ -0,0 +1,72 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    private final static RowMapper<Reservation> RESERVATION_ROW_MAPPER =
```
mapper 분리는 어떠한 장점을 갖는지 궁금해요.
아래 query에서 column명 변경이 되면 같이 변경되어야 할텐데 더 잘 관리할 수 없을까요?

yuyoungrhee replied:
**mapper 분리 장점**
- 책임의 분리
    - sql문을 통해 query를 날리는 로직 / 결과를 Reservation 객체로 매핑하는 책임
- 가독성 향상

**column명 변경 문제**
- as 절로 컬럼명을 rowMapper와 맞출 수 있음 -> column 명이 변경 되어도 rowMapper에는 영향 없음 
- 참고: 현재는 as 절이 일부에만 적용되어있습니다! 추후 반영 가능할 것 같습니다

단점은 충분히 커버할 수 있고, 분리로써 얻을 수 있는 장점이 있기 때문에 분리하게 되었습니다.

dusdn1702 commented on src/main/java/roomescape/repository/JdbcReservationRepository.java:
```diff
@@ -0,0 +1,72 @@
+                    t.id as time_id,
+                    t.start_at as time_value
+                FROM reservation as r
+                inner join reservation_time as t
```
inner join 은 어떤 이유에서 사용하셨나요?

yuyoungrhee replied:
Reservation은 필드로 ReservationTime을 가지고 있기 때문에 온전한 Reservation을 구성하려면 timeId를 통한 ReservationTIme의 start_at 컬럼이 필요합니다. 

start_at 컬럼을 위해서는 db를 한 번 더 조회할 수도 있지만, Inner join을 하여 1번의 db 접근으로 Reservation 객체를 구성할 수 있도록 하였습니다.

dusdn1702 replied:
join은 사용하면 안됐나요?? 🤔

yuyoungrhee replied:
앗 join도 사용할 수 있을 것 같습니다. 기본 JOIN == INNER JOIN 이라고 하네요..! 이 부분은 확인하지 못했습니다. 

LEFT JOIN이나 RIGHT JOIN 등 다양한 JOIN이 활용되지 않는 사용이라면 간단하게 JOIN을 사용하는 것이 좋은 것 같습니다. 기능적인 차이는 없는 것 같아 이번에는 JOIN으로 바꿔보도록 하겠습니다!(코드가 간단한게 현재는 더 좋은 것 같아서요👍)

[refactor: as절 추가, inner join -> join 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/0419ec5881c662bc77b4628af570fcd0df6fdc91)

dusdn1702 commented on src/main/java/roomescape/repository/JdbcReservationRepository.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public Long save(Reservation reservation) {
+        SimpleJdbcInsert simpleJdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
```
find에선 query, save에선 SimpleJdbcInsert을 선택하신 이유가 무엇인가요?

yuyoungrhee replied:
SimpleJdbcInsert는 Insert 전용 도구이며,
테이블 이름, 컬럼, 키를 지정하면 자동으로 insert 쿼리를 구성해주는 간편함을 제공합니다!

이런 기능은 insert 한정으로 제공되기 때문에 나머지 쿼리에는 사용하지 않았습니다.

insert는 db 측에서 필요한 컬럼이 정해져있고, 항상 동일하게 작성하게 됩니다. 따라서 패턴화와 자동화가 가능하기 때문에 제공되는 기능입니다. 저 또한 실수를 줄이고 간편하게 코드를 작성하기 위해 사용해보았습니다.

dusdn1702 commented on src/main/java/roomescape/service/TimeService.java:
```diff
@@ -0,0 +1,50 @@
+        return TimeResponse.toDto(ReservationTime.withId(id, reservationTime));
+    }
+
+    private void validateRequestTime(String time) {
```
메서드 순서는 어떻게 가져가는 것이 좋을까요?

yuyoungrhee replied:
접근 제어자 기준과, 수직 거리 기반 정렬 2가지가 있을 것 같습니다. 

**접근 제어자 기준(public - protected - private)**
- 외부 api 파악이 쉬움(클래스 사용자 입장에서 좋음)
    - 즉 이 클래스를 직접 사용해야하는 입장에서 좋다고 생각

**수직 거리 기반**
- 코드 읽기가 편하고 이해가 쉬움
    - 해당 코드가 어떤 역할을 하는지 코드 순서대로 쭉 알 수 있음

TimeService 같은 경우 수직 거리를 고려한 코드입니다. validateRequestTime은 여러 메서드에서 공유되는 util이 아니라 사용되는 메서드가 정확히 존재하므로 수직거리를 고려하는 것이 더 좋다고 판단했습니다. 만약 여러 메서드에서 재사용 되는 경우, 가장 아래로 배치했을 것 같습니다.

dusdn1702 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,51 @@
+        return ReservationMapper.toDto(Reservation.withId(id, reservation));
+    }
+
+    private void validateDateFormat(String date) {
```
`LocalDate.parse(date);` 파싱이 불필요하게 중복되고 있는데 이 부분 개선해보면 좋겠어요.

yuyoungrhee replied:
Request(dto)에서 **문자열에서의 날짜 형식 검증**만 가능하다고 생각하여 이런 식으로 구현했는데, 날짜 형식 또한 `@JsonFormat` 을 통하여 검증 가능하다는 것을 알았습니다.

불필요하게 중복된다면 parseDate 등으로 바로 return 하는 형식으로 바꾸어 사용할 수 있겠습니다. 현재는 해당 날짜 검증 로직을 `ReservationRequest`로 옮기면서 관련 로직이 모두 삭제됐습니다.

``` java
public record ReservationRequest(

        @NotNull
        @JsonFormat(pattern = "yyyy-MM-dd")
        LocalDate date,

        @NotBlank
        String name,

        @NotNull
        Long timeId
) {
}
```

dusdn1702 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,51 @@
+        return ReservationMapper.toDtos(reservations);
+    }
+
+    public void cancelReservation(Long id) {
```
cancel 인가요? 사용자로선 삭제했던 것 같은데요 ㅎㅎ

yuyoungrhee replied:
controller와 service의 메서드명이 똑같은 것에 의문을 느끼고 변경해보는 시도를 해보았습니다..!

관련해서 공식적인 내용이나 신빙성 있는 자료를 찾을 수는 없었습니다. 관례에 가까워서 명시적으로 정의된 부분이 없었기에 찾기가 어려웠습니다.. 우선 제 생각을 말씀드리겠습니다.

**Controller**
- Http요청을 수행한다는 의미를 더 나타냄
- create, get, update, delete 등

**Service**
- 메서드가 수행하는 비즈니스 로직적 의미를 더 담아냄
- cancel, search, modify 등

이러한 이유로 Controller와 Service가 메서드에 담고싶은 의미가 우연히 같다면 같은 메서드명을 사용했지만, delete -> cancel에 한해서는 변경해보았습니다. 어떻게 생각하시나요?

dusdn1702 replied:
비즈니스 로직 상 삭제가 아니라 취소라고 생각하신 이유가 무엇인가요?
비슷한 의미이지만 실제 취소하는 요구사항이 추가된다면 명명을 어떻게 하실 것 같으신가요?
http 메서드, query, 사용자 노출 문구 모두 삭제여서 취소가 어색했네요 ㅎㅎ

yuyoungrhee replied:
이 부분은 제가 잘못 생각하고 있었던 것 같습니다. 자연스럽게 "예약자"의 관점에서 바라보고 있었네요. 
이번 방탈출 예약 관리 프로그램은 **관리자 입장입니다**.. 관리자의 입장에서는 당연 "취소" 가아니라 "삭제" 가 맞겠네요. 수정했습니다🙇‍♂️

[refactor(ReservationService): 예약을 삭제하는 메서드명 cancel -> delete 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/57e6fcc2eb09ce86233ec6eb92754c07dab27d80)

dusdn1702 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,51 @@
+        Reservation reservation = ReservationMapper.toDomain(request, reservationTime);
+        Long id = reservationRepository.save(reservation);
+
+        return ReservationMapper.toDto(Reservation.withId(id, reservation));
```
예약 생성 후에 id를 반환하고 있는데요, 여기서 다시 withId를 통해 Reservation을 생성해야할 이유가 있을까요?

yuyoungrhee replied:
saved된 "도메인" 기준으로 응답을 반환해야 신뢰성이 있다고 생각했기 때문입니다.

request와 생성 후 반환되는 id 만으로도 ReservationResponse를 생성할 수 있지만, 아직 Service에서 비즈니스 로직을 거치기 전인 request를 id와 함께 그대로 다시 반환해주는 것은 응답에 대한 신뢰도가 떨어진다고 생각했습니다.

추가로 ReservationRepository에서 id가 아닌 id가 부여된 객체를 반환받는 방법도 생각해봤습니다.
```java
ex) Reservationsaved = reservationRepository.save(reservation);
```
 그러나 Repository에서 도메인의 정적펙터리 메서드를 사용한다는 부분에서, 서비스보다 Repository에 생성 책임을 두는 것이 더 좋은 이유를 찾지 못했습니다.

제가 고려하지 못한 부분이 있을까요? 성능적인 부분에서 불필요한 객체 생성이라고 판단하시는 걸까요?

dusdn1702 replied:
신뢰성을 위한다면 save한 후에 find로 해당 값을 다시 찾아와서 반환하는게 가장 신뢰도가 높지 않을까요?
만일 return 전에 어떤 사유로 name이 변경됐다면 변경된 name이 노출되어야하지 않을까요?
성능적인 이슈가 되진 않지만 의도가 궁금했어요.

yuyoungrhee replied:
이 부분에 대해서 다시 생각해보았습니다..

1. Request + id -> Response 구성
2. id가 부여된 Domain -> Response 구성
3. Repository에 저장된 Domain -> Response 구성

**1번과 2번**
이렇게 되어있을 때, 답변 당시 1번보다 2번이 신뢰도가 높다고 했지만, 사실상 느낌만 그런 것이고 신뢰도는 동일하다고 생각이 바뀌었습니다. 실제로 같은 정보입니다.

그러나 코드에서 주려는 뉘앙스가 미묘하게 다르다고 생각하는데요, 바로 Reservation이라는 새로운 불변객체를 통해서 Response를 생성한다는 것입니다. 
**코드에 담고 싶은 의도: 저장 후에는 이런 객체가 되었고, 객체가 Response로 매핑되어서 반환해준다.**

1, 2번의 단점: 만약 저장 도중 DB에서 필드가 수정되는 경우 응답에 대한 신뢰도가 깨질 것 같습니다. 

**3번**
3번은 확실하게 신뢰도가 높습니다.
단점: 성능상 findById 한 번은 차이가 별로 없다고 해도, DB에 다시 한 번 접근하는 비용 및 코드 복잡도가 추가됩니다. 저장 직후의 응답에 대한 신뢰도가 정말 중요한 상황이라면 3번을 고려해볼 것 같습니다. 

결론: 지금 상태에서 "DB에있는 정보가 이것이다"가 아닌, "도메인이 이렇게 구성되었다"를 전달해주는 것으로 **우선 충분하다고 생각**

dusdn1702 replied:
좋습니다 👍 
앞으로 프로젝트의 요구사항이 다양해지면서 상황에 따라 변경하는 것으로 이해했어요.
서버 개발자로서 db를 어떻게 관리하고 어떻게 응답을 내릴지에 대해 계속 고민하게 될텐데 어디에 집중해야할지 계속 고민해보면 좋겠어요 😄

dusdn1702 commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,45 @@
+        this.reservationTime = reservationTime;
+    }
+
+    public static Reservation of(Long id, String name, LocalDate date, ReservationTime time) {
```
각각의 Reservation을 생성하는 static 메서드들의 기준이 무엇인가요? 추후 관리에 용이한 방법이라고 생각하시나요?

yuyoungrhee replied:
```java
    public static Reservation of(Long id, String name, LocalDate date, ReservationTime time) {
        return new Reservation(id, name, date, time);
    }

    public static Reservation assignId(Long id, Reservation reservation) {
        return new Reservation(id, reservation.getName(), reservation.getReservationDate(), reservation.getReservationTime());
    }

    public static Reservation withoutId(String name, LocalDate reservationDate, ReservationTime reservationTime) {
        return new Reservation(null, name, reservationDate, reservationTime);
    }
```

`of`: 모든 필드를 받아 생성하므로 가장 기본적이라는 의미로 관습적인 네이밍인 `of`를 사용했습니다.

`withoutId`, `assignId`: 메서드명을 통해 어떻게 생성하고자 하는지 의미를 부여하기 위해 정적 펙터리 메서드를 사용했습니다. 데이터베이스에 저장되기 전에는 id가 null이고, 저장된 후에 id가 부여되므로 `Reservation` 객체에서 생성 방법을 제공해주는 방법을 선택했습니다.

메서드명으로 하고자하는 바가 충분히 전달된다고 생각하여 관리가 용이하다 생각되는데, 어떤 포인트에서 코멘트 주신 걸까요? 

추가로 Reservation에서 id가 null인 상태를 만들어내는 것이 이상적이라고 생각하진 않지만, id가없는 Dto 혹은 Vo 코드의 추가가 현재 수준에서는 복잡도만 더 증가시킨다고 생각하여 고려하지 않았습니다.

dusdn1702 replied:
of 하나만 존재해도 된다고 생각했어요. 
assignId, withoutId 모두 단 하나의 상황에서 사용되는 메서드이기 때문이에요.
reservation 객체를 생성하는 상황이 추가된다면 또 새로운 팩토리 메서드를 생성하는걸 의도하신거죠?

yuyoungrhee replied:
일부 동의합니다. registerReservation 과정에서만 일회성으로 사용되고 있습니다. 

이렇게 설계한 이유는 객체가 어떤 상태에서 생성되었는지를 명시적으로 표현하고 싶었기 때문입니다.
- ID가 없는 상태를 생성하는 것(저장 전)
- ID를 부여하여 생성하는 것(저장 후)

of를 사용한다면 null을 외부에서 주게 될텐데, 그럼 id가 null인 객체를 생성한다는 의미는 외부에서 갖게 됩니다. 이 의미 부여를 Reservation 객체 안으로 가져오고 싶었습니다.

객체를 생성하는 상황이 추가된다면 새로운 팩토리 메서드를 추가한다기보다,
**생성의 의미를 표현해야될만큼 중요하다**면 새로운 팩토리 메서드를 생성하는 것을 의도하고 싶습니다만.. 어떻게 더 잘 표현할 수 있을지는 잘 모르겠네요🥲

dusdn1702 replied:
좋습니다 ㅎㅎ 계속 미션들 잔행하면서 디벨롭 시켜보면 될 것 같네요!

dusdn1702 commented on src/test/java/roomescape/dto/request/ReservationRequestTest.java:
```diff
@@ -0,0 +1,16 @@
+    @Test
+    void createRequestFail_when_invalidFormattedDate() {
+        String date = "20222-02-02";
+        //assertThatThrownBy(() -> new ReservationRequest(date, "멍구", 1L));
```
확인 부탁드려요~

yuyoungrhee replied:
Request에 대해서 간단한 검증을 임시적으로 해본다는 게 그대로 커밋되었네요..! 

Request 같은 경우 현재 jakarta.validation 라이브러리를 사용해 검증하는 것으로 변경되었습니다. 따라서 `@SpringTest`나 `@WebMvcTest`를 사용하는 ControllerTest가 필요합니다.

해당 검증의 경우 4-9단계의 요구사항(JUnit을 이용한 단위테스트에 집중)에서 벗어나는 것 같아 구현 삭제했습니다.

dusdn1702 commented on src/test/java/roomescape/testRepository/FakeReservationRepository.java:
```diff
@@ -0,0 +1,37 @@
+import roomescape.domain.Reservation;
+import roomescape.repository.ReservationRepository;
+
+public class FakeReservationRepository implements ReservationRepository {
```
fake를 도입하신 이유가 궁금해요.
쿼리에 대한 검증은 어디서 되고 있을까요?

yuyoungrhee replied:
### Fake 도입 이유

fake는 Service 계층의 테스트를 위해 도입했습니다.

Service에서 쿼리를 직접 날리는 JdbcRepository를 그대로 테스트에서 사용하게 된다면 다음과 같은 문제가 있습니다.
1. 속도가 느림 
    - 실제로 db에 접근해야하므로 매우 느려짐
2. repository에 과도하게 의존함
    - DB 상태에 따라 결과가 달라짐
    - 서비스의 목적은 "비즈니스 로직이 잘 동작하는지"인데, DB가 제대로 동작하는지까지 검증하게 됨 -> 단위테스트가 아님
    - "Repository는 제대로 돌아간다고 믿고" 테스트해야하는 문제

따라서 간단한 인메모리 로직으로 대체되는 FakeRepository를 구현해서 테스트에 사용했습니다.

### test double의 다른 방법들
Service를 테스트하는 다른 방법으로는 대표적으로 Stub/Mock 이 있을 것 같습니다. 
- Stub: 미리 정의된 값을 리턴 
- Mock: 호출 여부 등을 검증
- Stub/Mock의 단점: **상태를 기억할 수 없음**
- 라이브러리 의존적

- Spy: 실제로 JdbcRepository를 호출하므로 속도 면에서 이점 없음, 호출 여부만 검증
- Dummy: 아예 동작하지 않는 객체이므로 무의미. (서비스 테스트에 넘겨야하지만 실제로 사용되지 않는 의존성 같은 경우에 사용 가능)


테스트의 신뢰성을 위해 응답 전체(결과 및 상태)를 검증하고 싶었고 Fake를 선택했습니다.

Fake의 단점으로는, Fake도 결국 사람이 짜는 코드이기 때문에 신뢰하기 어렵다는 문제가 있을 것 같습니다. 구현이 가볍지 않고 복잡도가 있다면, FakeRepository조차 테스트해야할 것입니다.

현재는 단순히 List에 add, remove, filter로 찾기 정도를 구현하기 때문에 오작동 가능성이 거의 없고, 명확하다고 생각합니다. 그럼에도 필요하다면 FakeRepository를 위한 테스트를 추가할 의향이 있습니다.

### 쿼리에 대한 검증 (Repository 테스트)
쿼리에 대한 검증은 없습니다. `@JdbcTest` 등으로 관련 의존성을 등록해야하기 때문에 순수 JUnit으로 테스트가 불가능한 영역입니다. 

새로운 테스트 도구나 기법을 도입하지 말라는 요구사항에 따라 현재 집중해야할 영역이 아니라고 생각했습니다. 충분히 리팩터링 된 후에 추가할 예정입니다. 현재는 제공되는 MissionStepTest로 간접적으로 통합테스트 검증이 되고있다고 생각합니다.

dusdn1702 replied:
좋습니다.ㅎㅎ
쿼리 검증도 `@JdbcTest` 추가가 아닌 MissionStepTest에서 사용한 `@SpringBootTest`로 시도해봐도 좋을 것 같아요.

yuyoungrhee replied:
`@SpringBootTest`와 `@JdbcTest` 둘 다 시도해보았습니다.

JdbcTest: JDBC 관련 빈만 등록합니다. (jdbcTemplate, DataSource 등) SpringBootTest: 모든 빈을 등록합니다(Service부터 Controller 등등 애플리케이션을 실제로 띄우기 위한 모든 빈)

Repository만 검증하는 테스트라면, 필요한 빈만 가볍게 등록하는 @JdbcTest가 더 적합하다고 판단했습니다.
(불필요한 빈을 띄울 필요가 없고, 테스트 속도도 더 빠릅니다.)

그런데 현재 테스트가 많지 않아서 그런지, 성능 차이가 거의 나지 않았습니다. 따라서 현재 테스트 수준에서는 상관 없다고 생각합니다만 필요없는 빈까지 등록해줄 필요는 없다고 생각되어 `@JdbcTest`를 활용해보았습니다.. 

[test(repository): jdbc repository 테스트 추가 및 MissionStepTest 일부 수정](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/defbf5b4370d12c5320622fb3c993e3a0608d0d9)

혹시 `@SpringBootTest`로 시도해보라는 조언은 학습 경험 차원이었을까요? 아니면 제가 놓친, `@SpringBootTest`가 더 알맞은 이유가 있을까요?

dusdn1702 replied:
오히려 말씀 주신 것처럼 `@JdbcTest`를 반드시 사용할 필요가 없기 때문에 지금 학습하지 않아도 될 것 같아서 기존에 사용했던 `@SpringBootTest`로 작성해봐도 될 것 같다는 생각에 말씀 드린 부분이었어요 ㅎㅎ
더욱이 비교 학습까지 해주셔서 감사합니다! 👍

dusdn1702 reviewed (changes requested):
안녕하세요 멍구~
DB를 도입하는 미션도 잘 수행해주셨네요. 고생 많으셨습니다.
코멘트 확인 부탁드릴게요!

dusdn1702 commented on src/test/java/roomescape/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,136 @@
+
+    @DisplayName("잘못된 형식의 date로 예약 등록 요청했을 때 예외가 발생한다.")
+    @ParameterizedTest
+    @ValueSource(strings = {"20000-04-01", "2024-04", "2024,10,00"})
```
꼼꼼한 테스트 좋네요 👍

yuyoungrhee committed:
refactor(ReservationTime): id를 부여하는 정적펙터리메서드명 변경 withId -> assignId

yuyoungrhee committed:
refactor(dto): 날짜나 시간 요청 형식에 대한 검증을 request에서 하도록 검증 위치 변경

yuyoungrhee committed:
refactor(mapper): ReservationTime에 관한 dto <-> domain 변환 로직을 책임지는 mapper 구현

yuyoungrhee commented:
샐리 꼼꼼한 리뷰 감사합니다! 코멘트에 대한 답변 위주로 진행했어요.

코드 변경사항은 많지 않은데, 크게 2가지입니다.

1. dto에서 String이 아닌 LocalDate / LocalTime과 같은 타입을 가지도록 변경
    - dto에서 바로 LocalDate / LocalTime이라는 유효한 타입을 가질 수 있도록 했습니다.
    - 원하는 형태로 요청 받기 또는 응답 보내기를 하려면 String 형태로 가공해야한다고 생각했는데, `@JsonFormat`을 활용하여 가공 문제를 해결했습니다.

2. ReservationTime에 관한 dto(request, response)에서도 dto <-> domain 변환 로직을 담당하는 `ReservationTimeMapper` 구현
    - [코멘트](https://github.com/woowacourse/spring-roomescape-admin/pull/288#discussion_r2058340078)와 같은 생각으로, 일관성을 위해 변경하였습니다.
    
`+` 변경을 빼먹어서 api Response 형식이 맞지 않는 부분 추가 커밋했습니다..!!! 앞으로는 미리 확인하겠습니다🙌🏽

yuyoungrhee commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,39 @@
+        this.timeService = timeService;
+    }
+
+    public ReservationResponse registerReservation(ReservationRequest request) {
```
Service가 Request DTO에 의존하고 있습니다.
Request는 View로부터 받은 값을 담는 Presentation 계층의 객체라고 생각해서,
Service가 이를 직접 받는 건 레이어드 아키텍처에서 말하는
“의존성은 상위(Controller) → 하위(Service)로 흐른다”는 원칙을 깨뜨리는 것 아닌가? 라는 고민이 생겼습니다.

그럼에도 불구하고 제 입장은,
도메인으로 변환하는 책임은 Controller가 아닌 Service에 있어야 한다고 생각합니다.
이 변환 자체도 비즈니스 로직의 일부이기 때문입니다.

이런 구조가 레이어드 아키텍처를 어긴다고 봐야 할지, 이 정도는 허용되는 타협이라고 봐도 되는 건지 궁금해졌습니다. 여러 논쟁(?)들을 찾아본 결과 보통 request를 그대로 Service에 전달해주면서 타협을 보는 거 같은데, 그렇다면 레이어드 아키텍처는 엄격하게 지키기에는 어려운 구조인가..같은 의문이 생겼습니다.

지금 상태는 어쩔 수 없구나~ 정도로 납득되었는데, 이에 관련해서 간단하게라도 샐리의 생각이 궁금하여 코멘트 남깁니다!!

dusdn1702 replied:
저도 많이 고민했던 부분인 것 같아요.
모든 아키텍처는 트레이드 오프가 존재하고, 그 안에서 선택을 하는 것은 개발자의 몫이죠.
멍구만의 기준을 만드는 것이 중요하다고 생각해요. 엄격한 아키텍처를 지키면서 얻는 것은 무엇이라고 생각하시나요?
판단하기에 변환이 비즈니스 로직의 일부라면 service에서 변환을 하는 것이 맞다고 보여요 👍

yuyoungrhee replied:
### **레이어드 아키텍처의 원칙**
레이어드 아키텍처가 원래 추구하려고 했던 점은 "의존성이 상위에서 하위로만 흐른다"입니다.
이는 상위의 변경이 하위 계층에 영향을 미치지 않도록 함입니다.

현재 문제: presentation 계층이 변경된다면 현재 service가 영향을 받습니다.

따라서 엄격하게 계층을 나누고 싶다면, controller용 requestDto/responseDto와 service용 requestDto/responseDto를 나눠야합니다.

### **코드 복잡도 vs 엄격한 아키텍처 원칙 준수**
현재 요구사항 수준에서는 controller용 == service용 이라고 보았습니다. 따라서 코드 복잡도를 줄이고 같은 dto를 전달받게 했습니다.

최대한 단순하게 설계하고, 필요할 때 분리해보고 싶습니다! dto를 나누는 것은 어려운 일이 아니라고 생각해서요. 확장성을 고려해서 설계하는 것도 중요하지만 현재 필요한 코드만을 작성하는 것도 중요하다고 생각합니다..!

dusdn1702 replied:
좋습니다! 의도를 정확히 갖고, 일관성 있는 코드를 작성하면 문제되는 것이 없다고 생각해요 👍

yuyoungrhee committed:
fix(response): api명세와 형식이 다른 부분 @JsonFormat으로 수정

dusdn1702 commented on src/main/java/roomescape/dto/request/ReservationRequest.java:
```diff
@@ -1,8 +1,20 @@
 package roomescape.dto.request;
```
패키지는 어떠한 기준을 갖고 나누셨는지 궁금해요!

yuyoungrhee replied:
큰 의식이 없었습니다. 그저 같은 기능 별로만 우선 패키징했던 것 같습니다.
레이어드아키텍처를 적용했기 때문에 계층 별로 패키지가 나누어져있으면 좋을 것 같습니다.

따라서 dto를 controller 패키지로 옮겼습니다(현재 request, response는 외부 요청, 응답을 표현하는 controller의 dto이므로)

[refactor: 패지키 구조 변경(dto, mapper)](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/44a102e96316b806c736ac85d484174d58f1a583)

dusdn1702 commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,45 @@
+import java.time.LocalDate;
+
+public class Reservation {
+    private final Long id;
```
도메인에서 db의 id를 가지고 있는 부분에 대해서 어떻게 보시나요?

yuyoungrhee replied:
**Reservation의 도메인이 id라는 필드를 가지고있는 것은 DB와 상관없이 당연한 일이라고 생각했습니다.** 

1. 요구사항에 따라 예약 기록을 관리해야합니다.
2. Reservation의 필드만으로는 식별할 수 없습니다. 중복된 예약은 다른 예약으로 바라봅니다.
3. 따라서 식별하기 위해 id 필드가 필요하게 되었습니다.
4. id라는 필드 자체도, domain의 속성이 됩니다.

장기같은 경우 좌표가 식별자 역할을 충분히 할 수 있기 때문에 id가 필요 없었습니다. 현재 예약 관리의 경우 Reservation의 name, date, time 필드만으로 식별할 수 없다고 판단했기 때문에 식별자를 만들게 되었습니다.

**Reservation에 id를 부여하는 방식이 2가지가 있습니다.**

1. Reservation을 생성할 때 id 부여 ->  DB에 그대로 저장
    - 장점: 
        - 도메인 독립성: DB에 의존하지 않고 id를 자체 관리
        - 객체 생성시 id 포함 모든 필드가 존재함->도메인을 도메인 자체로 믿을 수 있음
    - 단점
        - 복잡성 증가: id 생성 로직을 비즈니스로직 어딘가에서 추가해야 함
        
3. Reservation을 id 없이 생성 -> DB에 저장(db의 auto increment 사용) -> db에서 부여된 id를 Reservation에서 부여
    - 장점
        - 이미 db가 잘 제공해주는 auto increment등의 id 자동 생성 기능 활용 가능
        - 코드가 간단해짐
    - 단점
        - 도메인은 저장 전까지 id가 없는 상황 존재
        - DB에 종속됨, DB 없이 id 부여 불가


저는 여기서 2번을 선택했기 때문에 db에 종속적인 도메인이 되었습니다. 그러나 현재로서는 그대로 유지하는 게 문제가 크지 않다고 생각합니다.

문제가 되는 곳은 Service의 `registerReservation()` 메서드 입니다. 이 곳에서 id가 null인 Reservation이 잠깐 존재하게 됩니다. 때문에 이 메서드 안에서 누군가 id가 null인 객체를 꺼내서 이것저것 하려고 한다면 정말 큰 문제이겠으나..! 이런 점까지 고려를 해서 코드의 복잡도를 늘려야될까에 대한 고민이 있어 최종적으로 i**d를 db에 저장하여 부여하는 것으로 유지하게 되었습니다.**
- 누군가 열어둔 getter를 사용해서 코드를 다 망쳐버리면 어떡하지? 누군가 이 줄을 실수로 삭제해버리면 어떡하지? 와 같은 수준의 고민이라고 생각했습니다....

dusdn1702 replied:
> Reservation의 도메인이 id라는 필드를 가지고있는 것은 DB와 상관없이 당연한 일이라고 생각했습니다.

동일한 예약이 존재할 수 있어 식별을 위해 id가 필요하기 때문에 도메인에 포함되어야 한다는 의견에 크게 공감되네요 좋습니다 😄

dusdn1702 commented on src/main/java/roomescape/dto/response/ReservationResponse.java:
```diff
@@ -1,9 +1,18 @@
+
 public record ReservationResponse(
+
         Long id,
```
primitive를 사용하지 않은 이유가 궁금해요!

yuyoungrhee replied:
그저 도메인의 타입을 그대로 반영했습니다. 생각해본 적 없는 내용이네요.. 관련 내용을 찾아보고 생각을 정리해보았습니다!

primitive 타입은 Jackson 변환 과정에서 null을 처리할 수 없어 예외가 발생합니다.
반면 Wrapper 타입은 null을 그대로 응답으로 전달할 수 있습니다.

성능 이슈는 부차적인 문제라고 생각했고, 가장 중요한 것은 “응답에 null을 허용할 것인가” 라는 의도라고 보았습니다.

이번 경우 응답 시 id에 null을 허용하고 싶지 않았습니다. Reservation에는 db 저장 전에만 null이 존재하고 무조건 id가 부여되어있어야하기 때문입니다. 만약 null이 들어온다면 서버에서 예외를 발생시켜야 한다고 판단했습니다.
따라서 primitive 타입(long)으로 변경하였습니다.

[refactor(response): id를 primitive 타입(long)으로 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/db1f422fea146b24a6a5a21cb453a5733fa37fe6)

---

코드에 의미를 담으며 작성하려고 했는데, 의식의 흐름대로 따라가다보니 제 생각은 담겨있지 않은 코드들도 많이 작성되는 것 같네요. 이번 코멘트로 알게 되었습니다.. 앞으로는 좀 더 의식적인 코드 작성을 연습해보려고 합니다🤩

dusdn1702 replied:
request와 domain에도 id가 모두 존재하고 Wrapper type으로 구현되어 있는데, 이 값들은 null로 들어가는 상황이 있기 때문에 그냥 두신 건가요??

dusdn1702 commented on src/main/java/roomescape/mapper/ReservationMapper.java:
```diff
@@ -1,33 +1,24 @@
-        LocalDateTime reservationTime = LocalDateTime.of(date, time);
 
-        return Reservation.createReservationWithoutID(request.name(), reservationTime);
+    public static Reservation toDomain(ReservationRequest request, ReservationTime reservationTime) {
```
mapper는 request->domain, domain->response 모두 수행을 하는데요, service에서 수행하는건 어때요?

yuyoungrhee replied:
변환도 하나의 책임이라고 생각해서 클래스를 분리해주고 싶었습니다.

다만 패키지 구조를 바꾸면서 service.mapper로 **service에서 사용되는 mapper**라는 의미를 더 나타내보았습니다..!

[refactor: 패지키 구조 변경(dto, mapper)](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/44a102e96316b806c736ac85d484174d58f1a583)

dusdn1702 reviewed (changes requested):
안녕하세요 멍구!
피드백 코멘트 확인했습니다 ㅎㅎ
꼼꼼하게 고민하고 작성해주셔서 감사해요.
코멘트 답변 위주로 더 이야기 나누고 싶은 부분들 추가했어요.
확인 부탁드릴게요!

yuyoungrhee committed:
refactor: as절 추가, inner join -> join 변경

- 코드 간결성을 위해 inner join -> join 변경(기능 같음)
- rowMapper와 조회된 컬럼명을 항상 일치지시키 위해 as절 활용

yuyoungrhee committed:
refactor(ReservationService): 예약을 삭제하는 메서드명 cancel -> delete 변경

yuyoungrhee committed:
refactor(response): id를 primitive 타입(long)으로 변경

- 응답으로 id=Null을 허용하지 않겠다는 의도

yuyoungrhee committed:
test(repository): jdbc repository 테스트 추가 및 MissionStepTest 일부 수정

- MissonStepTest: 각 테스트가 독립적으로 동작하도록 수정

yuyoungrhee committed:
refactor: 패지키 구조 변경(dto, mapper)

yuyoungrhee committed:
test(MissonStepTest): @DirtiesContext 삭제

@BeforeEach에서의 db 초기화가 이미 이루어지고 있으므로 삭제함

yuyoungrhee commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -45,6 +45,13 @@ public class MissionStepTest {
     @BeforeEach
     void setUp() {
         RestAssured.port = port;
+
+        jdbcTemplate.execute("SET REFERENTIAL_INTEGRITY FALSE");
+        jdbcTemplate.execute("TRUNCATE TABLE reservation");
+        jdbcTemplate.execute("ALTER TABLE reservation ALTER COLUMN id RESTART WITH 1");
+        jdbcTemplate.execute("TRUNCATE TABLE reservation_time");
+        jdbcTemplate.execute("ALTER TABLE reservation_time ALTER COLUMN id RESTART WITH 1");
+        jdbcTemplate.execute("SET REFERENTIAL_INTEGRITY TRUE");
```
겪었던 문제: TRUNCATE TABLE을 했음에도 id가 1부터가 아닌 더 높은 값부터 생성되는 문제 발생

테스트 독립성을 위해 각 테스트 전 데이터를 초기화해주는 작업을 추가했습니다.
그러나 h2는 TRUNCATE TABLE을 실행하더라도  AUTO_INCREMENT 컬럼 값이 초기값으로 재설정되지 않는다는 문제가 있었습니다.

때문에 다른 테스트에서 2개를 삽입했다면, TRUNCATE TABLE을 실행해도 다른 테스트에서는 id가 3부터 시작하는 문제가 발생합니다. 이에 따라 테스트에서 id를 직접 지정하는 과정, 외래키 제약 조건 등에서 예외가 발생하게 되었습니다.(사용하려는 id가 없거나 이미 존재)

해결: TRUNCATE TABLE + id 1로 초기화

`+` @BeforeEach의 Repository clear 로 인해 더이상 @DirtiesContext를 쓸 필요가 없다는 사실을 알고 추후 제거하는 작업도 진행했습니다.

[test(MissonStepTest): @DirtiesContext 삭제](https://github.com/woowacourse/spring-roomescape-admin/pull/288/commits/61848f778169bb9d9017e193707322efb0da7f43)

yuyoungrhee commented on src/test/java/roomescape/repository/JdbcReservationRepositoryTest.java:
```diff
@@ -0,0 +1,107 @@
+
+@JdbcTest
+@Import(JdbcReservationRepository.class)
+@ActiveProfiles("test")
```
**test용 h2를 분리하여 실행되도록 했습니다.**
현재는 프로덕션 코드 또한 애플리케이션이 종료되면 데이터가 삭제되지만, 만약 향후 데이터 유지가 필요해지더라도 test로 인한 프로덕션 데이터 변경이 없도록 분리했습니다.

MissionStepTest에서는 test Profile을 적용하지 않았습니다. `connection.getCatalog()` 으로 DB 이름을 가져오는 검증이 4단계에 있었는데, testDB의 이름이 변경되면서 테스트가 실패하게 됩니다. MissionStepTest의 검증 코드는 최대한 건드리고싶지 않아서 냅두었습니다..!
(test Proife을 사용하고, TEST_DB로 기대값을 맞추게 된다면 잘 작동합니다)

dusdn1702 replied:
좋습니다 👍👍👍

yuyoungrhee commented:
생각할만한 주제를 많이 주셔서 감사합니다..! 코멘트 모두 달아두었습니다.

주요 변경 사항
- 패키지 구조 변경: controller/dto service/mapper
- jdbc Repository 테스트 추가
- MissionStepTest 테스트 독립화, @DirtiesContext 제거&각 테스트 전 Repository clear
- Response에서의 id 타입 primitive type(long)으로 변경
- sql inner join -> join 단순화

dusdn1702 commented on src/main/java/roomescape/controller/dto/request/ReservationRequest.java:
```diff
@@ -0,0 +1,20 @@
+package roomescape.controller.dto.request;
```
dto 패키지의 위치를 변경해주셨는데, service 단의 mapper에서도 사용되는 만큼 위치를 다시 한번 생각해보면 좋겠어요

dusdn1702 reviewed (approved):
안녕하세요 멍구~
피드백에 대해 하나하나 깊이 있게 고민하고, 코멘트로 작성해주셔서 감사합니다. 고생 많으셨습니다.
코드적으로 더 수정할 부분은 없어보여 이번 미션은 여기서 마무리하도록 할게요.
코멘트 몇가지 남긴 부분들은 확인 한번 해주시면 좋을 것 같아요.
앞으로도 화이팅입니다 😄
